### PR TITLE
JIT: Add pred block iterator that tolerates pred list modifications

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Follow-up to #99362. `fgRedirectTargetEdge` modifies the predecessor lists of the old and new successor blocks, so if we want to be able to use it in `fgReplaceJumpTarget`, we need a pred block iterator that is resilient to these changes, as we frequently call the latter method while iterating predecessors.

I haven't found the need to update the pred edge iterator to allow updates too, but I imagine it would look the same as the pred block iterator.

`fgTailMergeThrows` looks a bit awkward right now. After this change is merged, I plan to add a new method similar to `fgRedirectTargetEdge` for `BBJ_COND` blocks, and leverage it in `fgReplaceJumpTarget`. Since that new method will preserve edge weights, we won't lose anything in `fgTailMergeThrows` by just calling `fgReplaceJumpTarget`. I think that will be a good opportunity to clean up `fgTailMergeThrows` by getting rid of all its helpers for updating edges.

cc @dotnet/jit-contrib, @AndyAyersMS PTAL.

### CODE REVIEW

File: src/coreclr/jit/block.h [hunk 1]
```diff
@@ -2399,28 +2401,43 @@ inline PredEdgeList::iterator& PredEdgeList::iterator::operator++()
...
 
 #ifdef DEBUG
-    // Check that the next block is the one we expect to see.
-    assert(next == m_next);
-    m_next = (next == nullptr) ? nullptr : next->getNextPredEdge();
+    // If allowEdits=false, check that the next block is the one we expect to see.
+    // If allowEdits=true, the user can modify the predecessor list while traversing it,
+    // so (next == m_next) may not be true. Since we cached the next pointer in m_next, this won't break iteration.
+    assert((next == m_next) || allowEdits);
+    updateNextPointer = true;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/block.h:2432]: This seems to be modifying live state under debug, which I think we should avoid.

**Review by AndyAyersMS — COMMENTED** [review]
I like this direction, but would prefer that the nature of the iterator be obvious from its name..

Maybe something like:

```C++
class PredBlockList
{
public:

   static PredBlockList PredBlocks() const
   {
      return PredBlockList(bbPreds, /*allowEdits*/ false);
   }

   static PredBlockList PredBlocksEditing() const
   {
      return PredBlockList(bbPreds, /*allowEdits*/ true);
   }

private:

   PredBlockList(FlowEdge* list, bool allowEdits) {...};
};
```

**Comment by amanasifkhalid** [comment]
@AndyAyersMS I separated out the `allowEdits=true` path in `operator++`. It has a bit of duplicated code now, but I think it makes it clearer that `m_next` is always updated for editing iterators, but only updated for non-editing iterators in Debug, so there shouldn't be any newly-introduced cases where we update state in Debug only. I also added `PredBlocksEditing`, but kept the templated pattern with the hope that we can avoid some branches in the iterator logic -- are you ok with this approach?

**Review by AndyAyersMS — APPROVED** [review]
This looks great.

**Comment by amanasifkhalid** [comment]
Thanks for the reviews! I should note that this change has [small diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=596983&view=ms.vss-build-web.run-extensions-tab) from slight changes in edge weights.

**Pull request merged**